README.md
# flutter_authorize_net_client

[![pub package](https://img.shields.io/pub/v/flutter_authorize_net_client.svg)](https://pub.dev/packages/flutter_authorize_net_client)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)

This is a simple flutter plugin allow using [authorize.net](https://authorize.net) using its unique API approach.

This plugin is implemented using very simple methods & models to integrate the authorize.net API.

Please refer to the developer api references for more details:
https://developer.authorize.net/api/reference

**Note**:

Only a limited set of operations are supported for now.
Also, all the transactions are Credit card based transactions only.


### Supported Operations:
- Testing Authentication of the client using credentials
    - Reference: https://developer.authorize.net/api/reference/index.html#gettingstarted-section-section-header
- Charge a credit card
    - Reference: https://developer.authorize.net/api/reference/index.html#payment-transactions-charge-a-credit-card
- Create an authorized payment for future transaction
    - Reference: https://developer.authorize.net/api/reference/index.html#payment-transactions-authorize-a-credit-card
- execute pre-authorized payment transaction
    - Reference: https://developer.authorize.net/api/reference/index.html#payment-transactions-capture-funds-authorized-through-another-channel
- void an existing authorized transaction
    - Reference: https://developer.authorize.net/api/reference/index.html#payment-transactions-void-a-transaction

### Install plugin:
Visit https://pub.dev/packages/flutter_authorize_net_client#-installing-tab-

### Import library:
```dart
import 'package:flutter_authorize_net_client/flutter_authorize_net_client.dart';
```

## Usage Examples:

### Create Client instance:
```dart
final _client = AuthorizeNetClient(
                  '5KP3u95bQpv',
                  '346HZ32z3fP4hTG2',
                  // provide AuthorizeNetClient.ENV_PRODUCTION here to use production api
                  environment: AuthorizeNetClient.ENV_TEST,
                );
```

### Test Authentication:
```dart

    final response = await _client.authenticationTest();
    print('isSuccessFul: ${response.isSuccessful}');

```

### Charge a credit card:
```dart

    final response = await _client.chargeCreditCard(
      '5',
      'USD'.toLowerCase(),
      '5424000000000015',
      '2022-12',
      '123',
    );
    print('isSuccessFul: ${response.isSuccessful}');
```

### Pre-Authorize payment:
```dart

    final response = await _client.authorizeCardPayment(
        '5',
        'USD'.toLowerCase(),
        '5424000000000015',
        '2022-12',
        '123',
    );

    print('isSuccessFul: ${response.isSuccessful}');
    // keep a reference to transactionID in memory for future use
    _refID = response?.transactionResponse?.transId;

```

### Execute Pre-Authorize payment:
```dart

    final response = await _client.priorAuthCaptureTransaction(
        '5',
        'USD'.toLowerCase(),
        '5424000000000015',
        '2022-12',
        '123',
    );

   print('isSuccessFul: ${response.isSuccessful}');
    // keep a reference to transactionID in memory for future use
    _refID = response?.transactionResponse?.transId;

```

### Void payment:
```dart

    final response = await _client.voidTransaction(_refID);
    print('isSuccessFul: ${response.isSuccessful}');

```


**Some notes regarding API parameters**:
- Currency must be in [ISO-4217](https://en.wikipedia.org/wiki/ISO_4217) format
    -  Value must be passed in lowercase,
	i.e `usd` instead of `USD`, `gbp` instead of `GBP`, `eur` instead of `EUR`, etc
    - Currency symbols are not supported
- Amount can be a decimal value
    - Up to 15 digits with a decimal point
    - This is the total amount and must include tax, shipping, tips, and any other charges.

## Contribution/Support
- File an issue on the repository, if something is not working as expected.
   - Please follow the issue template used in flutter-sdk's repository, may be we'll integrate that here as well.
- File an issue in the repository, If you have any suggestions and/or feature requests, use `[Suggestion]` or `[FeatureRequest]` tags in issue titles.
- To support you just have to help out fellow developers on of the filed issues in this repository.
- To contribute, just follow the standard open source contributions instructions, maybe we can follow the ones used in the flutter sdk. We'll see how it goes.



CHANGELOG.md
## [0.1.0+2] - 05/05/2021
#### Documentation update
* Added example code in method documentation
* Removed debug logs

*Note*: The dependencies are kept old to support projects that has not yet migrated to Null-safety
Will migrate to null safety in near future.


## [0.1.0+1] - 25/04/2021
#### Documentation update

## [0.1.0] - 25/04/2021
#### initial release.
Supported Operations:
- Testing Authentication of the client using credentials
- Charge a credit card
- Create an authorized payment for future transaction
- execute pre-authorized payment transaction
- void an existing authorized transaction


lib/flutter_authorize_net_client.dart
library flutter_authorize_net_client;

import 'dart:convert';

import 'src/models/models.dart';
import 'src/utils/network_utils.dart';
import 'src/utils/utils.dart';

/// Class that represents AuthorizeNet Client
///
/// This class provides an interface to interact with Authorize.net
/// API with ease
///
/// The requests & responses are converted to a models based interaction, instead
/// of plain json which is the input & output of the API
///
/// Supported Operations:
/// - Testing Authentication of the client using credentials
/// - Charge a credit card
/// - Create an authorized payment for future transaction
/// - execute pre-authorized payment transaction
/// - void an existing authorized transaction
///
///
/// Note: This class implements Singleton pattern, i.e only one instance of the
/// AuthorizeNetClient will be created for whole app
class AuthorizeNetClient {
  /// String value that represent Production environment
  static const ENV_PRODUCTION = 'production';

  /// String value that represent testing environment
  static const ENV_TEST = 'sandbox';

  /// String value which represents current environment of the client
  final String environment;

  /// String value which represents merchant name(API ID) of the client
  final String merchantName;

  /// String value which represents transaction key(API SECRET) of the client
  final String transactionKey;

  /// Internal constructor
  AuthorizeNetClient._(
    this.merchantName,
    this.transactionKey,
    this.environment,
  );

  /// static instance to implement singleton pattern
  static AuthorizeNetClient _instance;

  /// factory method to check & instantiate AuthorizeNetClient instance only once
  factory AuthorizeNetClient(
    String merchantName,
    String transactionKey, {
    String environment = ENV_TEST,
  }) {
    if (_instance == null) {
      _instance = AuthorizeNetClient._(
        merchantName,
        transactionKey,
        environment,
      );
    }
    return _instance;
  }

  /// Getter for api url, value is based on the current type of [environment]
  String get baseApi => environment == ENV_PRODUCTION
      ? 'https://api.authorize.net/xml/v1/request.api'
      : 'https://apitest.authorize.net/xml/v1/request.api';

  /// A getter to facilitate instantiating a [MerchantAuthentication] object
  /// when needed
  MerchantAuthentication get merchantAuthentication =>
      MerchantAuthentication(merchantName, transactionKey);

  /// -------- API Methods -------- ///

  /// Method to test current configuration of the client using provided
  /// credentials & environment
  ///
  /// Response of this method is useful to know if the current configuration
  /// is working properly or not
  ///
  /// If not then it will provide some error information to resolve the issue
  ///
  ///
  /// Example with test data:
  ///```dart
  ///   final response = await _client.authenticationTest();
  ///   print('isSuccessFul: ${response.isSuccessful}');
  /// ```
  ///
  /// https://developer.authorize.net/api/reference/index.html#gettingstarted-section-section-header
  Future<AuthenticationTestResponse> authenticationTest() async {
    final authenticationTestRequest =
        AuthenticationTestRequest(merchantAuthentication);

    var responseJson = await executeRequest(
      baseApi,
      authenticationTestRequest.getRequestJson(),
    );
    final response =
        AuthenticationTestResponse.fromJson(jsonDecode(responseJson));
    return response;
  }

  /// A Method to execute a transaction using the provided credit card details
  ///
  /// This method will use the provide card data, i.e. [cardNumber], [expirationDate] & [cardCode]
  /// & [amount], [currencyCode] to create a transaction request that will charge the
  /// given credit card the provided [amount]
  ///
  /// An optional [referenceID] can be provided to track & verify the transactions
  ///
  /// Example with test data:
  ///```dart
  ///   final response = await _client.chargeCreditCard(
  ///       '5',
  ///       'USD'.toLowerCase(), // converting to lowercase is important
  ///       '5424000000000015',
  ///       '2022-12',
  ///       '123',
  ///     );
  ///   print('isSuccessFul: ${response.isSuccessful}');
  /// ```
  ///
  /// **Note**:
  ///  Make sure that the provided amount includes all tax, shipping, duty etc charges
  ///
  /// Reference: https://developer.authorize.net/api/reference/index.html#gettingstarted-section-section-header
  Future<CreateTransactionResponse> chargeCreditCard(
    String amount,
    String currencyCode,
    String cardNumber,
    String expirationDate,
    String cardCode, {
    String referenceID,
  }) async {
    final transactionRequest = TransactionRequest.authCaptureTransaction(amount,
        currencyCode, Payment.creditCard(cardNumber, expirationDate, cardCode));
    final createTransactionRequest = CreateTransactionRequest(
      merchantAuthentication,
      referenceID: referenceID ?? uniqueReferenceID,
      transactionRequest: transactionRequest,
    );

    var responseJson = await executeRequest(
      baseApi,
      createTransactionRequest.getRequestJson(),
    );
    final response =
        CreateTransactionResponse.fromJson(jsonDecode(responseJson));
    return response;
  }

  /// A Method to Pre-Authorize a transaction to execute in future
  ///
  /// This method will use the provide card data, i.e. [cardNumber], [expirationDate] & [cardCode]
  /// & [amount], [currencyCode] to create a transaction request that will generate
  /// a transactionID value that can be used in future to execute a transaction.
  ///
  /// An optional [referenceID] can be provided to track & verify the transactions
  ///
  /// Example with test data:
  ///```dart
  ///   final response = await _client.authorizeCardPayment(
  ///       '5',
  ///       'USD'.toLowerCase(), // converting to lowercase is important
  ///       '5424000000000015',
  ///       '2022-12',
  ///       '123',
  ///     );
  ///   print('isSuccessFul: ${response.isSuccessful}');
  /// ```
  ///
  /// Reference: https://developer.authorize.net/api/reference/index.html#payment-transactions-authorize-a-credit-card
  Future<CreateTransactionResponse> authorizeCardPayment(
    String amount,
    String currencyCode,
    String cardNumber,
    String expirationDate,
    String cardCode, {
    String referenceID,
  }) async {
    final transactionRequest = TransactionRequest.authOnlyTransaction(amount,
        currencyCode, Payment.creditCard(cardNumber, expirationDate, cardCode));
    final createTransactionRequest = CreateTransactionRequest(
      merchantAuthentication,
      referenceID: referenceID ?? uniqueReferenceID,
      transactionRequest: transactionRequest,
    );

    var responseJson = await executeRequest(
      baseApi,
      createTransactionRequest.getRequestJson(),
    );
    final response =
        CreateTransactionResponse.fromJson(jsonDecode(responseJson));
    return response;
  }

  /// A Method to execute Pre-Authorized transaction
  ///
  ///
  /// This method will use the provide transaction data, i.e. [amount], [currencyCode]
  /// & [referenceTransactionID] to execute a pre-authorized transaction request that will charge the
  /// credit card used for pre-authorization
  ///
  /// An optional [referenceID] can be provided to track & verify the transactions
  ///
  ///
  /// Example with test data:
  ///```dart
  ///   final response = await _client.priorAuthCaptureTransaction(
  ///      '5',
  ///      'USD'.toLowerCase(),
  ///      _refID,
  ///     );
  ///   print('isSuccessFul: ${response.isSuccessful}');
  /// ```
  ///
  /// Reference: https://developer.authorize.net/api/reference/index.html#payment-transactions-capture-funds-authorized-through-another-channel
  Future<CreateTransactionResponse> priorAuthCaptureTransaction(
    String amount,
    String currencyCode,
    String referenceTransactionID, {
    String referenceID,
  }) async {
    final transactionRequest = TransactionRequest.priorAuthCaptureTransaction(
      amount,
      currencyCode,
      referenceTransactionID,
    );
    final createTransactionRequest = CreateTransactionRequest(
      merchantAuthentication,
      referenceID: referenceID ?? uniqueReferenceID,
      transactionRequest: transactionRequest,
    );

    var responseJson = await executeRequest(
      baseApi,
      createTransactionRequest.getRequestJson(),
    );
    final response =
        CreateTransactionResponse.fromJson(jsonDecode(responseJson));
    return response;
  }

  /// A Method to Void/Discard a Pre-Authorized transaction
  ///
  ///
  /// This method will void a pre-authorization transaction using the  [referenceTransactionID]
  ///
  /// After the transaction is void, it can not be used to charge the credit card used
  /// for the pre-authorization.
  ///
  /// A new transaction must happen in order to charge the credit card.
  ///
  /// An optional [referenceID] can be provided to track & verify the transactions
  ///
  /// Example with test data:
  ///```dart
  ///   final response = await _client.voidTransaction(_refID);
  ///   print('isSuccessFul: ${response.isSuccessful}');
  /// ```
  ///
  /// Reference: https://developer.authorize.net/api/reference/index.html#payment-transactions-void-a-transaction
  Future<CreateTransactionResponse> voidTransaction(
    String referenceTransactionID, {
    String referenceID,
  }) async {
    final transactionRequest = TransactionRequest.voidTransaction(
      referenceTransactionID,
    );
    final createTransactionRequest = CreateTransactionRequest(
      merchantAuthentication,
      referenceID: referenceID ?? uniqueReferenceID,
      transactionRequest: transactionRequest,
    );

    var responseJson = await executeRequest(
      baseApi,
      createTransactionRequest.getRequestJson(),
    );
    final response =
        CreateTransactionResponse.fromJson(jsonDecode(responseJson));
    return response;
  }
}
